embassy-nrf/src/low_power_uarte.rs

```diff
@@ -0,0 +1,302 @@
+//! Async low power UARTE.
+//!
+//! The peripheral is autmatically enabled and disabled as required to save power.
+//! Lowest power consumption can only be guaranteed if the send receive futures
+//! are dropped correctly (e.g. not using `mem::forget()`).
+
+use core::future::Future;
+use core::sync::atomic::{compiler_fence, Ordering};
+use core::task::{Context, Poll};
+
+use embassy::util::Signal;
+use embedded_dma::{StaticReadBuffer, StaticWriteBuffer};
+
+use crate::fmt::{assert, trace};
+use crate::hal::gpio::{Floating, Input, Output, Pin, PushPull};
+use crate::hal::pac;
+use crate::hal::prelude::*;
+use crate::hal::target_constants::EASY_DMA_SIZE;
+use crate::interrupt;
+
+// Re-export SVD variants to allow user to directly set values.
+pub use pac::uarte0::{baudrate::BAUDRATE_A as Baudrate, config::PARITY_A as Parity};
+
+/// Interface to the UARTE peripheral
+pub struct LowPowerUarte {
+    instance: pac::UARTE0,
+}
+
+struct State {
+    tx_done: Signal<()>,
+    rx_done: Signal<u32>,
+}
+
+static STATE: State = State {
+    tx_done: Signal::new(),
+    rx_done: Signal::new(),
+};
+
+impl LowPowerUarte {
+    pub fn new(uarte: pac::UARTE0) -> Self {
+        // Enable interrupts
+        uarte.events_endtx.reset();
+        uarte.events_endrx.reset();
+        uarte
+            .intenset
+            .write(|w| w.endtx().set().txstopped().set().endrx().set().rxto().set());
+        // TODO: Set interrupt priority?
+        interrupt::unpend(interrupt::UARTE0_UART0);
+        interrupt::enable(interrupt::UARTE0_UART0);
+
+        LowPowerUarte { instance: uarte }
+    }
+
+    /// Sets the baudrate, parity and assigns the pins to the UARTE peripheral.
+    // TODO: Make it take the same `Pins` structs nrf-hal (with optional RTS/CTS).
+    // TODO: #[cfg()] for smaller device variants without port register (nrf52810, ...).
+    pub fn configure(
+        &mut self,
+        rxd: &Pin<Input<Floating>>,
+        txd: &mut Pin<Output<PushPull>>,
+        parity: Parity,
+        baudrate: Baudrate,
+    ) {
```

Comment:
It'd be cool to take the same `Pins` struct as `BufferedUarte`. Perhaps moving `BufferedUarte` to the `uarte` module? so that the module structure is `uarte::Uarte, uarte::BufferedUarte, uarte::Pins`

Reply:
Agreed. I changed the method to take the nrf-hal Pin structure. That one can also be used by `BufferedUarte`.